lighthouse/classes/event_properties/interfaces.py

```diff
@@ -0,0 +1,225 @@
+    in its interface class (EventPropertyInterface).
+    If a more advanced process is needed (like accessing an external service),
+    then the methods of this class should be rewritten (or a new class
+    should be created).
+    This class also contains methods that will be needed for internal
+    process for the class but should not be called from outside of this
+    instance (they should be considered all protected methods).
+    The public interface that should be used in this case is
+    defined in EventPropertyInterface.
+    """
+
+    def __init__(self, params: Dict[str, Any]):
+        """
+        Stores the params, resets the instance and perform a validation check on the params
+
+        Arguments:
+            params: Dict[str, Any] - dict of key/value with the settings for
+            this property
+        """
+        self.reset()
+        self._params = params
+        if self._params is None:
+            raise ValidationError("You need to define params to create the EventProperty")
+
+    def reset(self):
+        """
+        Deletes all errors and clears the validation setting.
+        """
+        self._errors: List[str] = []
+        self._value = None
+        self._validate = True
+
+    @property
+    def errors(self) -> List[str]:
+        """
+        Validates the instance and returns the complete list of errors found (this
+        errors are not just validations; it could also be previous exceptions
+        thrown during the lifetime of this instance).
+
+        Arguments:
+            None
+
+        Returns:
+            List[str] - List of error messages found currently
+        """
+        self.validate()
+        return self._errors
+
+    def valid(self) -> bool:
+        """Alias for #validate()"""
+
+        return self.validate()
+
+    def enforce_validation(self) -> None:
+        """
+        Raises a ValidationError exception if the instance does not pass validation.
+
+        Returns:
+            ValidationError - Raises exception
+        """
+        if not self.validate():
+            raise ValidationError("Validation error")
+
+    def process_validation(self, condition: bool, message: str) -> None:
+        """
+        Stores the error message if the condition is not True.
+        Changes the validation state for the instance.
+        It is recommended to run this inside a validation_scope() context
+        to be able to properly log any possible exceptions while checking the
+        condition.
+
+        Arguments:
+            condition: bool - Condition to check
+            message: str - Error message to store
+
+        Returns:
+            None
+
+        """
+        if not condition:
+            if message not in self._errors:
+                self._errors.append(message)
+            self._validate = False
+        self._validate = self._validate and True
+
+    @contextmanager
+    def validation_scope(self):
```

Comment:
It creates a try: catch: that will grab any exceptions and logging them as error inside the instance without raising the exception further.
This is use in the validation process to be able to check several conditions checks, but if one of then raise an exception it won't block us to continue checking the next conditions checks; so we can have a complete list of all errors.

Reply:
The retrieval_scope() down is similar in the way that is 'unsafe', we want to logging the exception as an error inside this class, but we dont' want to block it (so it is rereaised); that way, we have the error message for this in this instance in case we need it. Also it marks the instance as invalid (so any subsequent call to it is_valid will be not valid), and any subsequent call to value will need to be reprocessed.

Reply:
I added a state machine about how the validation flag would be affected by the different situations of the event property: 
https://lucid.app/lucidchart/de929982-0693-44cd-8680-e2d563808c93/edit?page=0_0#